middleman_protocol/middleman_protocol/message.py

```diff
@@ -0,0 +1,368 @@
+    - parsing serialized bytes payload back to original format on deserialization,
+    - validation of original payload before serialization.
+
+    """
+
+    __slots__ = [
+        'payload',
+        'payload_type',
+        'request_id',
+    ]
+
+    def __init__(self, payload: Any, request_id: int) -> None:
+        self._validate_request_id(request_id)
+        self.request_id = request_id
+
+        self._validate_payload(payload)
+        self.payload = payload
+
+    @classmethod
+    @abstractclassmethod
+    def _deserialize_payload(cls, payload: bytes):
+        pass
+
+    @abstractmethod
+    def _serialize_payload(self, payload: Any):
+        pass
+
+    @abstractmethod
+    def _validate_payload(self, payload: Any):
+        pass
+
+    @classmethod
+    def get_frame_format(cls) -> Struct:
+        """ Returns Struct object containing frame structure. """
+        frame_format = Struct(
+            frame_signature=Bytes(FRAME_SIGNATURE_BYTES_LENGTH),
+            signed_part_of_the_frame=Struct(
+                request_id=BytesInteger(FRAME_REQUEST_ID_BYTES_LENGTH),
+                payload_type=Enum(Byte, PayloadType),
+                payload_length=BytesInteger(FRAME_PAYLOAD_LENGTH_BYTES_LENGTH),
+                payload=GreedyBytes,
+            ),
+        )
+
+        assert FRAME_PAYLOAD_LENGTH_STARTING_BYTE == (
+            frame_format.frame_signature.subcon.length +
+            frame_format.signed_part_of_the_frame.request_id.subcon.length +
+            1  # 1 byte for payload_type
+        )
+        assert FRAME_PAYLOAD_STARTING_BYTE == (
+            frame_format.frame_signature.subcon.length +
+            frame_format.signed_part_of_the_frame.request_id.subcon.length +
+            1 +  # 1 byte for payload_type
+            frame_format.signed_part_of_the_frame.payload_length.subcon.length
+        )
+
+        return frame_format
+
+    @classmethod
+    def add_frame_separator(cls, raw_frame: bytes) -> bytes:
+        """ Adds frame separator to raw frame. """
+        assert isinstance(raw_frame, bytes)
+
+        return raw_frame + FRAME_SEPARATOR
+
+    @classmethod
+    def remove_frame_separator(cls, raw_message: bytes) -> bytes:
+        """ Remove frame separator from raw frame. """
+        assert isinstance(raw_message, bytes)
+        assert FRAME_SEPARATOR in raw_message
+
+        frame_separator_index = raw_message.index(FRAME_SEPARATOR)
+
+        assert len(raw_message) == frame_separator_index + FRAME_SEPARATOR_LENGTH
+
+        return raw_message[:frame_separator_index]
+
+    @classmethod
+    def deserialize(
+        cls,
+        raw_message: bytes,
+        public_key: bytes,
+        is_frame_separator_expected: bool=False,
+        is_encoding_expected: bool=False,
+    ) -> Any:
+        """
+        Parses and validates received message in bytes.
+        Returns original payload.
+        """
+
+        # Remove frame separator from raw message if it is expected
```

Comment:
`FRAME_SEPARATOR` itself shouldn't be escaped, only its occurrences in raw_message, to avoid situation when receiver thinks he already have full message before receiving actual separator.

Reply:
Ok, now I get it :)